examples/pybind11/external_usm_allocation/external_usm_allocation/_usm_alloc_example.cpp

```diff
@@ -120,6 +121,42 @@ py::list tolist(DMatrix &m)
     return rows;
 }
 
+class USMDeleter
+{
+public:
+    USMDeleter() = delete;
+    USMDeleter(const USMDeleter &) = default;
+    USMDeleter(USMDeleter &&) = default;
+    USMDeleter(const ::sycl::queue &queue) : _context(queue.get_context()) {}
+    USMDeleter(const ::sycl::context &context) : _context(context) {}
+
+    /*! @brief call operator to delete resources */
+    template <typename T> void operator()(T *ptr) const
+    {
+        try {
+            ::sycl::free(ptr, _context);
+        } catch (const std::exception &e) {
+            std::cout << "Call to sycl::free caught an exception: " << e.what()
+                      << std::endl;
+        }
+    }
+
+private:
+    ::sycl::context _context;
+};
+
+dpctl::memory::usm_memory make_zeroed_device_memory(size_t nbytes,
+                                                    sycl::queue &q)
+{
+    char *data = sycl::malloc_device<char>(nbytes, q);
+    q.memset(data, 0, nbytes).wait();
+
+    USMDeleter _deleter(q);
+    auto shptr = std::shared_ptr<void>(data, _deleter);
```

Comment:
I'm usually using the following construct for device memory:
```C++
template<class T> using DeviceMem = std::unique_ptr<T[], std::function<void (T*)>>;

template<class T> DeviceMem<T> malloc_device(int64_t count, cl::sycl::queue& queue)
{
    return DeviceMem<T>(sycl::malloc_device<T>(count, queue), [queue](T* mem)
    {
        sycl::free(mem, queue);
    });
}
```

So you don't need to create `USMDeleter` and replace it with lambda which is much less verbose.

Reply:
Thank you for the suggestion @AlexanderKalistratov . I have pushed a change to the example to use lambda with capture instead of `USMDeleter`.